### ISSUE TITLE
Allow creation of object instances by interface

### ISSUE THREAD

**Issue by andyward** [xBimTeam/XbimEssentials#360]
Just a random thought. Is it viable to do some kind of poor-mans IoC / DI implementation so that we can create new instances of entities in the correct schema purely by specifying their interface rather than the concrete type. 

That would make it much easier to create code targetting multiple schemas with the same code. Bearing in mind the `IModel` knows the Schemas version and has an Entity Factory, why *can't* we construct an entity like:

```csharp
var rel = model.Instances.New<IIfcRelContainedInSpatialStructure>();  // Using an interface not a Type
... 
```

... and have the ModelInstances sort out constructing the appropriate type

Based on this issue:

https://github.com/xBimTeam/XbimWindowsUI/issues/165

**Comment by martin1cerny** [comment]
We have this. 

```cs
// IFC2x3 or IFC4 model
var create = new Create(model);

// returns IIfcWall
var wall = create.Wall();
```

**Comment by bekraft** [comment]
Hi,

I also had a try on to implement such a feature. Being able to abstract the query structure / builder pattern apart from physical schema helps a lot. For simple case classes (i.e. IFC products which all have the same pattern) it works well. But I finally ended in implementing extension methods for special cases, since there are also differences between the IFC schemes (and logic). 

The most critical issue here is that you will need a standalone interface API as a common superset. Currently it's basically the most recent (Ifc4) interface abstraction, isn't it? So you have to pass i.e. IFC4 interfaces to the builder which than has to detect the proper implementation based on the physical model scheme. 

What's the plan for the upcoming IFC releases? Will IFC2x3 implement IFC4x3 interfaces - if applicable? What about dropped entities / removed from schema?

Cheers,
Bernold

**Comment by RichardWhitfield** [comment]
> We have this.
> 
> ```cs
> // IFC2x3 or IFC4 model
> var create = new Create(model);
> 
> // returns IIfcWall
> var wall = create.Wall();
> ```

Hi Martin, this may already have had some discussion so apologies if I'm rehashing something which is already in progress.  The issue with the proposed approach is that we already need a concrete version of the model because the interfaces are part of the IFC4 and IFC2x3 assemblies.

I think a solution, if possible, would be to lift the element interfaces to the `Xbim.Ifc` base assembly so that the types are able to be resolved via DI at runtime.  I've put some sample code below to illustrate the point.

```cs
using Xbim.Ifc;
using Xbim.IO;
using Xbim.Common;
using Xbim.Common.Step21;

public class SomeClass
{
    private readonly IEntityFactory _entityFactory;
    //Require IoC on creation of class with the IEntityFactory - injected as part of DI
    public SomeClass(IEntityFactory entityFactory)
    {
    _entityFactory = entityFactory;
    }
    
    public SomeMethod(IFormFile stepFile)
    {
    //To this point it is possible to use DI and use a different _entityFactory injected at runtime
    var model = new Xbim.IO.Memory.MemoryModel(_entityFactory);
    model.LoadStep21(stepFile.OpenReadStream(), stepFile.Length);
    
    //Here is where the issue arises...because the interface for IfcColumn is tied to the concrete implementation of the Ifc version 
    (ie Ifc4 or Ifc2x3) then it's no longer possible to continue with DI at this point.
    var columns = model.Instances.OfType<IIfcColumn>();
    }
```

The hard part about that it that you will need a *lot* of DI logic, ie essentially logic for every object.  I would think that there must be a better solution for this so I can have a look.

Let me know if you would like me to look into any of this.  Happy to run with a small example first to show the proposal and go from there.

Cheers, Rich

**Comment by RichardWhitfield** [comment]
Apologies for the multi posts.  I had an initial look at this and can report the following (which you might already be aware of).

1. What I was referring to as the 'base' `Ifc` assembly (ie the one with no number suffixes) actually uses both `Ifc2x3` and `Ifc4` assemblies.  As such I tried initially to move these interfaces to the `Xbim.Common` project.  
In an effort to come to the table with solutions and not problems, my suggestion for this would probably be to create a new project called `Xbim.IfcBase` which would give a place for these interfaces to live as well as any other shared logic.
*Because the existing implementation has interfaces in the implementations of the `Ifc4` and `Ifc2x3` assemblies this would be a breaking change so I don't know how you feel about this*.
2. The structs in the existing `Ifc4` and `Ifc2x3` implementations do implement an interface so we have a case where the interfaces (for example `IIfcActorRole` has a concrete attribute of type `IfcLabel`.
I think this is relatively straightforward and just needs a similar Interface for the structs to implement.  
*This should not be a breaking change*.
3. Based on the way that I chopped the code up there are some bad assumptions - for example I assumed that the Interface would always be before the concrete class in the file but this isn't always the case (for example `IfcActorSelect` does not conform to this assumption.
*No biggies, I just need better logic here.  Thought it worth getting some feedback first though.*

At the moment they seem to be the biggest items...I'd have to solve them to see anything else in the litany of compiler errors.  Thought it was a good place to stop and see what the general concensus was.

**Comment by martin1cerny** [comment]
Yes, implementing IFC from scratch is not exactly straightforward task :-) 

The use case you are talking about is already supported. If you write your controller against IFC4 interfaces, it will work for IFC2x3 and IFC4x1 as well. That is how we use it quite heavily ourselves.

**Comment by RichardWhitfield** [comment]
Thanks Martin.  That's great, I will go down that path.
